### ISSUE TITLE
`no-extra-semi`: False positive when required between symbols

### ISSUE THREAD

**Issue by adamlui** [eslint-stylistic/eslint-stylistic#622]
### Validations

- [X] Follow our [Code of Conduct](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://eslint.style/contribute/guide).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

### Describe the bug

For example here https://github.com/adamlui/ai-web-extensions/blob/main/chatgpt-infinity/chrome/extension/content.js#L275-L281:

```js
...
    };

    // Add STARS styles
    ['black', 'white'].forEach(color => document.head.append(
        dom.create.elem('link', { rel: 'stylesheet',
            href: `https://assets.aiwebextensions.com/styles/css/${color}-rising-stars.min.css?v=50f457d`
    })))
...
```

... it is required after the `}` or else the code is treated like `{...}[...]` and the script crashes

### Reproduction

https://github.com/adamlui/ai-web-extensions/actions/runs/12044843022/job/33582490607

### Contributes

- [X] I am willing to submit a PR to fix this issue
- [X] I am willing to submit a PR with failing tests

**Comment by 9romise** [comment]
It does not seems to be a bug.
I looked at all the failed cases in https://github.com/adamlui/ai-web-extensions/actions/runs/12044843022.
All of them are
```
if (boolean) {
  // ...some logic
};

['black', 'white'].forEach(() => {
  // ...some logic
})
```
Even if it doesn't have a semicolon after the if expression, it's still considered an `IfStatement` and an `ExpressionStatement`. Adding a semicolon brings an `EmptyStatement`, so this semicolon is indeed **extra**, as reported by this rule.

**Comment by adamlui** [comment]
@9romise it doesn't matter if it's an if or expression statement, the point is the semicolon is required or interpreter will read the code wrong and script crashes. Therefore the semicolon is absolutely required and not "extra" (this rule is supposed to safely warn users to remove them, removing here breaks script)

**Comment by adamlui** [comment]
@9romise look again at the code

```js
    if (!tweaksStyle || parseInt(tweaksStyle.getAttribute('last-updated')) < tweaksStyleUpdated) {
        if (!tweaksStyle) { // outright missing, create/id/attr/append it first
            tweaksStyle = dom.create.elem('style',
                { id: 'tweaks-style', 'last-updated': tweaksStyleUpdated.toString() })
            document.head.append(tweaksStyle)
        }
        tweaksStyle.innerText = '* { scrollbar-width: thin }' // make FF scrollbar skinny to not crop toggle
    }; // eslint-disable-line

    // Add STARS styles
    ['black', 'white'].forEach(color => document.head.append(
        dom.create.elem('link', { rel: 'stylesheet',
            href: `https://assets.aiwebextensions.com/styles/rising-stars/css/${color}.min.css?v=891df10`
    })))
```

Without required semicolon, interpreter reads it like `if (condition) { logic }[array].forEach(callback)` which is nonsensical, so it is not extra but required (has nothing to do with if/expressionempty statements, semicolons are sometimes required for correct interpretation and this rule must not interfere in those cases

**Comment by adamlui** [comment]
Another example to make you understand the point of linting rules: if programmer wrote `alert('hi') ; alert('there')` and a plugin warns it is "extra" user expects to remove it and code run fines. But just like in prev example, there will be uncaught error instead, because it is 100% **required** thus not an "extra" symbol then name of this rule implies what it's truly supposed to catch (i.e. not required ones like in both examples for script to run)

**Comment by adamlui** [comment]
Also the statement's not empty but even if it was, a `no-empty` warning is supposed to handle that. This rule is called `no-extra-semi` instead, means warn user don't use unrequired semis, not when it's required like here

**Comment by adamlui** [comment]
TLDR semicolon is required here (i.e. **not extra**) or else code breaks, so this bug must be patched to correctly account for curly brace at ends of if blocks and follow-up opening square brackets of arrays

(plugin already correctly accounts for other similar cases that do not show false warning like above, for example here:

```js
    update.style.tweaks() ; document.head.append(tweaksStyle);

    // Add STARS styles
    ['black', 'white'].forEach(color => document.head.append(
        dom.create.elem('link', { rel: 'stylesheet',
            href: `https://assets.aiwebextensions.com/styles/rising-stars/css/${color}.min.css?v=891df10`
    })))
```

**Comment by adamlui** [comment]
@9romise when you close this valid issue you're essentially telling user to work around the bug by doing whatever they're already doing (in this case using `eslint-disable-line`) which just makes code worse for having noise and making filesize larger, defeating purpose of ESLint

**Comment by 9romise** [comment]
The examples you give https://github.com/eslint-stylistic/eslint-stylistic/issues/622#issuecomment-2542688560 and https://github.com/eslint-stylistic/eslint-stylistic/issues/622#issuecomment-2542688560 don't match the issue, and semicolons are really needed here, but our issue is whether semicolons are needed after if expressions' `}` and before arrays' `[`.

Can you give a minimal repro like this?
```
if (condition) {
  // ...some logic
}
[array].forEach(() => {
  // ...some logic
})
```
I tried a simple use case but couldn't reproduce it, here are my use cases and results:
Windows Chrome 131.0.6778.140
![image](https://github.com/user-attachments/assets/2b071c2a-0d7c-4402-8eab-fccfb98b10fd)
Windows node.js 22.11.0
![image](https://github.com/user-attachments/assets/8fa39d58-37e1-4cfe-988f-f7ff23c7c35b)

**Comment by 9romise** [comment]
> @9romise when you close this valid issue you're essentially telling user to work around the bug by doing whatever they're already doing (in this case using `eslint-disable-line`) which just makes code worse for having noise and making filesize larger, defeating purpose of ESLint

I closed it only because I found out that it was a misjudgment, that it wasn't actually a bug, and that it didn't cause the error to happen (or that the existing information didn't allow me to reproduce it). If it's valid, I'd be happy to reopen it and fix it.

**Comment by adamlui** [comment]
@9romise I think you are right, in my example I tested it is not required! Sorry for my reaction

**Comment by adamlui** [comment]
But in the examples where it's truly required, ESLint doesn't warn user, do you agree it is good idea?

**Comment by adamlui** [comment]
For example paste this in console

```js
console.log('sup')
['adam', 'bob'].forEach(dude => console.log(dude))
```

... and without semi you get `Uncaught TypeError: console.log(...) is undefined`

**Comment by adamlui** [comment]
Also here https://github.com/adamlui/ai-web-extensions/blob/1c43a3e6a739ba06f50208968522b62c276e94b6/chatgpt-infinity/chrome/extension/content.js#L138-L148:

```js
...
    // Add LISTENER to auto-disable Infinity Mode
    document.onvisibilitychange = () => {
        if (config.infinityMode) {
            settings.save('infinityMode', false) ; syncConfigToUI({ updatedKey: 'infinityMode' }) }
    }

    // Add STARS styles
    ['black', 'white'].forEach(color => document.head.append(
        dom.create.elem('link', { rel: 'stylesheet',
            href: `https://assets.aiwebextensions.com/styles/rising-stars/css/${color}.min.css?v=891df10`
    })))
...
```

... if semicolon is inserted there is no `no-extra-semi` warning when there shuold be

**Comment by 9romise** [comment]
You can try [semi](https://eslint.style/rules/default/semi).

**Comment by adamlui** [comment]
> You can try using [semi](https://eslint.style/rules/default/semi).

> Rule Details
> 
> This rule enforces consistent use of semicolons.

But I and most others don't want to consistently use them, so isn't it a good idea to warn users when it's required as a general rule since code will break otherwise? (Not a stylistic warning but in case you have pull in the core project)

**Comment by 9romise** [comment]
I think it’s hard to say whether a semicolon is absolutely necessary since there are cases where it’s intentional.

For me, the `semi` and the [`beforeStatementContinuationChars`](https://eslint.style/rules/default/semi#beforestatementcontinuationchars) option in stylistic rules are good enough for handling this. Additionally, `indent` already shows unexpected indentation when a needed semicolon is missing, which works fine for me.

However, others might have better ideas, you could open an issue in the ESLint repo to bring more people into the discussion.

**Comment by adamlui** [comment]
> I think it’s hard to say whether a semicolon is absolutely necessary since there are cases where it’s intentional.

I meant this example:

```js
console.log('sup')
['adam', 'bob'].forEach(dude => console.log(dude))
```

... it is required to not break, so do you agree it is good idea in core ESLint to warn (not use stylistic `semi` since breaking is not stylistic reason)

> For me, the semi and the [beforeStatementContinuationChars](https://eslint.style/rules/default/semi#beforestatementcontinuationchars) option in stylistic rules are good enough for handling this. Additionally, indent already shows unexpected indentation when a needed semicolon is missing, which works fine for me.

But not everyone agrees on same styles like some don't want to import those rules at all, then the breaking change as in above example goes undetected

> However, others might have better ideas, you could open an issue in the ESLint repo to bring more people into the discussion.

Maybe I will but wanted to know if you agree first it is good idea to warn about missing semis that break code without relying on stylistic rules
